### DISCUSSION TITLE
Difference between Camera.defaultSettings and individual Camera props

### DISCUSSION THREAD

**Discussion by mattijsf** [rnmapbox/maps#3112] [Q&A] [answered]
In [this example](https://github.com/rnmapbox/maps/blob/7f166bb3e8d34e5ea163ee14c8ddab2e6b494807/example/src/examples/V10/Markers.tsx#L52-L56) I noticed the following Camera props:

```tsx
<Camera
  defaultSettings={{ centerCoordinate: centerCoord, zoomLevel: 14 }}
  centerCoordinate={centerCoord}
  zoomLevel={14}
/>
```

What is the main intention/difference between `defaultSettings` and individual `centerCoordinate` / `zoomLevel`-kinda props. 

Should the `defaultSettings` be seen as a "initial camera" prop? Meaning that changes this prop won't update the camera once the camera is updated once? Is there any benefit of using `defaultCamera` instead of the individual camera props?

**Comment by mfazekas** [comment, marked as answer]
@mattijsf yep, it's a non ideal solution, defaultSettings is the staring coordinates of the camera.  Kind-a of similar to RN TextInput's defaultValue. 

```jsx
<Camera
   defaultSettings={{ zoomLevel: 14}}
   zoomLevel={16}
/>
```


would start at zoom level 14 and  perform an animation to zoomLevel: 16. 

But you're right:

```jsx
<Camera
   defaultSettings={{zoomLevel: 14, ...}}
/>  
```

is same as 
```jsx
<Camera
   zoomLevel={14}
   ...
   animationDuration={0}
/>  
```

**Reply by mattijsf** [reply]
Thanks for clarifying. I'm trying to understand why my map still opens with the [center](https://docs.mapbox.com/style-spec/reference/root/#center) / [zoom](https://docs.mapbox.com/style-spec/reference/root/#zoom) present in the `styleURL` style json which is still visible for a _very_ short time after loading the map instead of the `defaultSettings` that I supply.

It's short but enough to experience some extra lag when opening the map.

**Reply by mfazekas** [reply]
Do you see it on iOS or Android?

**Comment by mattijsf** [comment]
I can only see it happening on Android. On iOS the map in general loads must faster so I can't really tell whether or not it could maybe be the same on a slower iPhone.

**Comment by mattijsf** [comment]
I just read about https://docs.mapbox.com/android/maps/guides/camera-and-animation/camera/#set-camera-on-map-initialization maybe we could try and see if we can expose `MapInitOptions.cameraOptions `. Is see we don't use this on both platforms. I'm not sure of props are available at this time of initialization though. What do you think?

Sorry for taking this discussion further. I could also make an issue so the topic is more focused.

**Reply by mfazekas** [reply]
So currently the init process looks like this:

```
RNMBXMapViewManger.createView => creates RNMBXMapView which creates MapView

RNMBXMapView .insertReactSubview(camera) adds the camera which then will execute the defaultStop (which comes from defaultSettings property)
```

I'll need to check android if there is anything we can do to improve with the current setup.
Otherwise we'd need to use MapInitOptions, and we have the following options there:
1.) Delay creating MapView until a Camera children is added (not sure if this is possible)
2.) In JS side check props.children for a Camera child and extract defaultSetting props for it and pass to MapView directly.
3.) Deprecate Camera#defaultSettings and add MapView#cameraDefaultSettings

**Reply by mattijsf** [reply]
1. Would make the initialization of Mapview (feel) slower. I already perceive it slower an pre-v10 so I wouldn't go for this (not necessarily due to this lib, could also be the mapbox SDK, I don't have numbers on this though).
2. Parsing children props is not great imo (anti-pattern?) but yeah API wise it might look better.
3. Makes most sense to me. Especially because you can have multiple Camera's in theory and behavior or each Camera.defaultSettings becomes more confusing.

I personally think using MapInitOptions for both platforms is the way to go as this would guarantee the most optimal circumstances for Mapbox to render the map at the desired region on style load. I'm afraid that every asynchronous method of updating the camera could have some negative side-effects on initialization resources (loading tiles / rendering etc). Difficult to prove though.

**Reply by mfazekas** [reply]
So actually on android we wait for StyleUrlLoaded to add camera. Which is necessary for layers but camera could be added earlier.

Before - I was able to capture this frame:
<img width="540" alt="image" src="https://github.com/rnmapbox/maps/assets/52435/9ecc7152-8db1-4a9c-bbf4-c774d8f4015f">

After - I've made a change so camera will be loaded before styleLoaded, and now I don't see frame showing location (0, 0)
https://github.com/rnmapbox/maps/assets/52435/5e4a74bf-9e3e-45c9-afb5-637270cd347d

**Reply by mattijsf** [reply]
Yeah looks good! So do I understand correctly that you just made this change?

**Reply by mfazekas** [reply]
See #3113
